static/js/components/forms/CouponForm.js

```diff
@@ -0,0 +1,298 @@
+// @flow
+import React from "react"
+import Picky from "react-picky"
+import DayPickerInput from "react-day-picker/DayPickerInput"
+import { Formik, Field, Form, ErrorMessage } from "formik"
+import * as yup from "yup"
+
+import type { Company, Product } from "../../flow/ecommerceTypes"
+
+type CouponFormProps = {
+  onSubmit: Function,
+  toggleForm: Function,
+  toggleProduct: Function,
+  isPromo: boolean,
+  companies: Array<Company>,
+  products: Array<Product>,
+  selectProducts: Function,
+  selectedProducts: Array<Product>,
+  productType: string
+}
+
+const baseValidations = {
+  name: yup
+    .string()
+    .required("Coupon name is required")
+    .matches(/^\w+$/, "Only letters, numbers, and underscores allowed"),
+  products:        yup.array().min(1, "${min} or more products must be selected"),
+  activation_date: yup.date().required("Valid activation date required"),
+  expiration_date: yup.date().required("Valid expiration date required"),
+  amount:          yup
+    .number()
+    .required("Percentage discount is required")
+    .min(1, "Must be at least ${min}")
+    .max(100, "Must be at most ${max}")
+}
+
+const promoValidations = yup.object(baseValidations).shape({
+  max_redemptions: yup
+    .number()
+    .min(1, "Must be at least ${min}")
+    .required("Number required"),
+  coupon_code: yup
+    .string()
+    .required("Coupon code is required")
+    .matches(/^\w+$/, "Only letters, numbers, and underscores allowed")
+})
+
+const singleUseValidations = yup.object(baseValidations).shape({
+  num_coupon_codes: yup
+    .number()
+    .min(1, "Must be at least ${min}")
+    .required("Number required"),
+  payment_transaction: yup.string().required("Payment transaction is required"),
+  payment_type:        yup.string().required("Payment type is required")
+})
+
+export const CouponForm = ({
+  onSubmit,
+  toggleForm,
+  toggleProduct,
+  isPromo,
+  companies,
+  products,
+  productType,
+  selectProducts,
+  selectedProducts
+}: CouponFormProps) => (
+  <Formik
+    onSubmit={onSubmit}
+    validationSchema={isPromo ? promoValidations : singleUseValidations}
```

Comment:
You can probably use `yup.when` instead so a change in form state doesn't cause the entire `Formik` component to re-render: https://github.com/jquense/yup#mixedwhenkeys-string--arraystring-builder-object--value-schema-schema-schema